collector/src/bin/collector.rs

```diff
@@ -2049,3 +2044,173 @@ fn get_commit_or_fake_it(sha: &str) -> anyhow::Result<Commit> {
+                    backends.push(CodegenBackend::Llvm);
+                }
+
+                // FIXME: remove this when/if NextArtifact::Commit's include/exclude
+                // changed from Option<String> to Vec<String>
+                // to not to manually parse args
+                let split_args = |l: Option<String>| -> Vec<String> {
+                    if let Some(l) = l {
+                        l.split(',').map(|arg| arg.trim().to_owned()).collect()
+                    } else {
+                        vec![]
+                    }
+                };
+                let sha = commit.sha.to_string();
+                let sysroot = Sysroot::install(sha.clone(), &target_triple, &backends)
+                    .with_context(|| format!("failed to install sysroot for {:?}", commit))?;
+
+                let mut benchmarks = get_compile_benchmarks(
+                    &compile_benchmark_dir,
+                    &split_args(include),
+                    &split_args(exclude),
+                    &[],
+                )?;
+                benchmarks.retain(|b| b.category().is_primary_or_secondary());
+
+                let artifact_id = ArtifactId::Commit(commit);
+                let mut conn = rt.block_on(pool.connection());
+                let toolchain = Toolchain::from_sysroot(&sysroot, sha);
+
+                let compile_config = CompileBenchmarkConfig {
+                    benchmarks,
+                    profiles: vec![Profile::Check, Profile::Debug, Profile::Doc, Profile::Opt],
+                    scenarios: Scenario::all(),
+                    backends,
+                    iterations: runs.map(|v| v as usize),
+                    is_self_profile: self_profile.self_profile,
+                    bench_rustc: bench_rustc.bench_rustc,
+                    targets: vec![Target::default()],
+                };
+                let runtime_suite = rt.block_on(load_runtime_benchmarks(
+                    conn.as_mut(),
+                    &runtime_benchmark_dir,
+                    CargoIsolationMode::Isolated,
+                    None,
+                    &toolchain,
+                    &artifact_id,
+                ))?;
+
+                let runtime_config = RuntimeBenchmarkConfig {
+                    runtime_suite,
+                    filter: BenchmarkFilter::keep_all(),
+                    iterations: DEFAULT_RUNTIME_ITERATIONS,
+                };
+                let shared = SharedBenchmarkConfig {
+                    artifact_id,
+                    toolchain,
+                };
+
+                run_benchmarks(
+                    &mut rt,
+                    conn,
+                    shared,
+                    Some(compile_config),
+                    Some(runtime_config),
+                )
+            }
+        }
+    });
+    // We need to send a message to this endpoint even if the collector panics
+    client.post(format!("{}/perf/onpush", site_url)).send()?;
```

Comment:
We shouldn't hit this endpoint from the queuing logic yet, because it could break the website if we hit the endpoint both from `BenchNext` and `BenchQueue`. I guess that only the benchmark logic could be extracted into a separate function, which will be called both by `BenchNext` and `BenchQueue`, but only `BenchNext` will trigger the endpoint.